dingo/core/utils/backward_compatibility.py

```diff
@@ -1,3 +1,80 @@
+from typing import Dict, List, Literal, Tuple
+
+import torch
+
+Device = Literal["meta", "cuda", "mps", "hip", "cpu"]
+
+
+def torch_available_devices() -> List[Device]:
+    """
+    Returns a list of all available PyTorch devices,
+    ordered: meta, cuda, mps, hip, cpu
+
+    Returns
+    -------
+    List of available device identifiers
+    """
+    devices: List[Device] = ["meta"]
+
+    # cuda
+    if torch.cuda.is_available():
+        devices.append("cuda")
+
+    # mps
+    try:
+        if torch.backends.mps.is_available():
+            devices.append("mps")
+    except AttributeError:
+        pass
+
+    # hip
+    try:
+        if hasattr(torch, "hip") and torch.hip.is_available():
+            devices.append("hip")
+    except AttributeError:
+        pass
+
+    # cpu
+    devices.append("cpu")
+
+    return devices
+
+
+def torch_load_with_fallback(
+    filename: str, preferred_map_location: Device = "cuda"
+) -> Tuple[Dict, Device]:
+    """
+    Loads a PyTorch file with fallback behavior:
+    1. Tries preferred_map_location (default: cuda)
+    2. Falls back to CUDA/MPS/HIP if available
+    3. Finally falls back to CPU
+
+    Returns
+    -------
+    Loaded model
+    """
+
+    try:
+        return (
+            torch.load(filename, map_location=preferred_map_location),
+            preferred_map_location,
+        )
+    except RuntimeError:
+        pass
+
+    devices = torch_available_devices()
+
+    for location in [d for d in devices if d != preferred_map_location]:
+        try:
+            return torch.load(filename, map_location=location), location
+        except RuntimeError:
+            pass
```

Comment:
Maybe one could add a print statement to indicate which device is being loaded? My suggestion would be: 

```suggestion
devices = torch_available_devices()
devices.insert(0, preferred_map_location)
for location in devices:
   try: 
      print(f"Attempting to load the model on {location}")
      return torch.load(filename, map_location=location), location
   except RuntimeError:
      print(f"Failed to load model on {location}, check your system settings")
      pass
```

That way if the situation Annalena mentioned happens (CPU training) the user would know in the logs and be able to debug from there.